saleor/payment/gateways/braintree/__init__.py

```diff
@@ -33,23 +31,31 @@
 }
 
 
-def get_customer_data(payment: Payment) -> Dict:
+class TransactionKind:
+    AUTH = 'auth'
+    CAPTURE = 'capture'
+    CHARGE = 'charge'
```

Comment:
But gateways can perform different operations. Some do not implement authorisation at all and instead do charge, some do not implement charge and require auth and capture. If you merge it, how will you tell what actually happened? And in case if one from multiple operations failed, which one?

Reply:
Does it matter though? Saleor says "auth and capture" and the backend does that, even if it takes several steps. Saleor says "auth but don't capture if possible" and the backend either does just auth or proceeds to capture anyway, there's no way around this. We can't really display "actually the store is configured to use a backend that does not implement preauth" to the user during checkout.

Reply:
Not in my eyes, no. There was a similar discussion here though https://github.com/mirumee/saleor/pull/3519#discussion_r245551018. The outcome is that if the gateway does auth and capture separately, then we want to know that. If it does only charge, then we want to know that as well. Although there's no "auth but do not capture" implemented. I still think this process_payment is a bad idea and I'd rather enforce `auth` and `capture` separately (if the gateway does not implement auth because it has charge only then who cares really? it can just do charge in capture step), but that was the outcome of the discussion.

Reply:
It's an implementation detail. Sure, we should generate a number of transaction items but there's no reason for either Saleor or the staff to care about implementation details of the payment process. A single store may have more than one payment method enabled and they can offer different sets of features. We can't say "oh, you can't pay after all" once the customer has entered their card details. Charging too early is an inconvenience (as in case of a cancelled order refund charges will apply) but it's still better than rejecting sales.